### ISSUE TITLE
[Persistence] Errors in persistence (after creating/modifying objects) should be visible to user

### ISSUE THREAD

**Issue by larkin** [nasa/openmctweb#58]
When there are errors that prevent an object from being persisted-- whether persisting a change to an object or persisting a new object-- the system should notify the user of the issue.

Currently, if a persistence operation fails, the interface acts like the persistence operation succeeded, and it's only when you refresh do you notice that changes have not been persisted.

We should not update the UI to show a persistence operation as completed until it has actually completed successfully, as subsequent actions performed while the UI is out of sync can cause significant negative effects.

**Comment by larkin** [comment]
Assigning to Victor, he will file a separate issue to design notifications around async operations and progress indicators, and he will implement this handling in different cases.

**Comment by charlesh88** [comment]
This need some UX design - Charles created issue #154.

**Comment by VWoeltjen** [comment]
Want to use infrastructure from #155 if possible, but that's a nice-to-have

**Comment by larkin** [comment]
Per tagup on 12/1/2015, narrow this down to complete in current sprint.

In persistence capability:
* handle false return from persistence service as promise rejection.
* throw global notification.

This leaves it open to implement per-action rejections at the persistence service so we have more information about errors and can provide errors that are more relevant to context.

**Comment by akhenry** [comment]
I require a little clarification here. So the persistence capability is (for now) going to produce a _blocking_ message indicating that there was an error, and reject the promise? Or should it be a non-blocking notification?

At some later date this will be finessed by removing the global error message, and implementing function-specific error messages.

**Comment by larkin** [comment]
is blocking vs unblocking something that is non-trivial to change?  Doesn't seem like it matters and ultimately it should be a tiny change to switch blocking to unblocking.

**Comment by akhenry** [comment]
Yeah it's totally trivial. Just wanted to confirm the approach, and that it's in the persistence capability that I'll make the change, and not further downstream.

**Comment by VWoeltjen** [comment]
Yep, persistence capability is the right place for this (insofar as it covers a broad range of cases.)

I believe we said blocking on the call earlier, so let's go with that for now since it will be easy to change later.

**Comment by akhenry** [comment]
So this ended up being a little more involve than I initially thought.

The persistence queue bundle has reasonably comprehensive persistence error reporting already. This takes the form of a blocking dialog that appears in most persistence failure modes (that I tried anyway). This actually appears after the PersistenceCapability is invoked so would result in duplication of dialogs if I were to implement a blocking dialog at that point. The existing functionality also provides the option to overwrite if the persistence error is due to a conflict in the remote store. As such, I'm reluctant at this point to remove this functionality

Moreover the CopyService has its own error handling using the DialogService.

As such, the approach I've gone for in the meantime is to implement failure messages as non-blocking notifications. This works alongside the existing blocking persistence notifications without being intrusive. It also provides error notification if queueing is disabled. If this approach sounds reasonable I'll add some unit tests and file a PR.

Longer term, there needs to be some reconciliation between the different approaches. If nothing else, the queueing errors need to look consistent with errors elsewhere. One option might be to integrate the newer notification options into the [PersistenceFailureHandler](https://github.com/nasa/openmctweb/blob/master/platform/persistence/queue/src/PersistenceFailureHandler.js). Another option is actually to refactor the PersistenceFailureHandler out of the queue bundle, and make it responsible for more generic persistence error handling, offering options such as overwrite, retry, etc.

**Comment by larkin** [comment]
I think we have an alternate issue for potentially removing the persistence queue; I think the persistence service doing anything more than resolve/reject is out of scope.  The service shouldn't have UI bindings.

I would expect individual operations to handle persistence failures differently as the failure of each user-action could potentially have different UX requirements, and each failure may require different clean up in the case of batch operations.  

At baseline: implement support for individual failures (e.g. copy fail, save fail, link fail) instead of a global persistence failure handler.  Once that is done, if we see a common pattern, then we support that, but from experience there isn't a global pattern that provides good UX without significant effort.

**Comment by akhenry** [comment]
I agree that the persistence errors should be bubbled up so that context appropriate handling can take place. The generic persistence error handler is proposed as an alternative that integrates with and extends the current persistence queue error handling.

There may be a case for generic handling of certain types of errors, or for abstracting details of the underlying persistence store. If we want to give users the option to overwrite or discard conflicting modifications for example, implementing this separately for copy, save, link, etc. doesn't make much sense. But in any case, any generic handling should happen higher in the stack.

**Comment by akhenry** [comment]
Have created #373 for addressing refactoring and reconciliation of differing approaches to persistence error notification.

**Comment by RCarvalhoNASA** [comment]
Testathon 20151207 - So there is a series of errors that appear in the bottom bar, but to my mind, they are a bit too subtle (small and dark at the bottom instead of big and bright in the center).  It still allows me to proceed around the tool even though the network is disconnected, and when I come back, that item still appears.  I'm not sure what behaviour would be enough to really make me notice, and know that the work I've done is not going to survive me closing the browser, but this doesn't seem to be enough.

**Comment by VWoeltjen** [comment]
I believe that switching to a blocking dialog should be fairly easy here. Still not ideal UX, but ensures the necessary information gets conveyed; we can improve iteratively from there. 

In nominal circumstances we don't expect to see these errors often (on console, connection to the server should be reliable; remotely, don't expect to create much) so this kind of messaging should not be practically disruptive

**Comment by charlesh88** [comment]
Testathon 20151207: Disconnected computer, tried to create an object. Notification appears, but once minimized overlay cannot be displayed while the system is disconnected. #383 filed to cover a fix to this behavior.

**Comment by VWoeltjen** [comment]
Quick fix for dialog not being available - prefetch all templates (e.g. during initialization of `mct-include` and `mct-representation` directives)

May need to prefetch earlier if this will compete with search indexing etc

**Comment by VWoeltjen** [comment]
Resolved by #388
